### PR TITLE
Issue is not overdue when it is on the same date #5566

### LINKED ISSUE #5566: Issues and Milestones due today are marked as overdue
- Gitea version (or commit ref): 1.6.1
- Git version:
- Operating system:
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Issues and Milestones can get an `Due date`.
If the due date is set to the current day it gets supplemented by the text `Overdue`.
In listings `Overdue` entries get their Due-Date colored red.

In fact [if something is overdue, it should have been done before now](https://www.macmillandictionary.com/dictionary/british/overdue) - if something is due today I still have time to do it 😄 

My suggestion would be to call something overdue if it really was due to yesterday or before.  
And I would suggest to add an additional state like `Due today` (which could be colored maybe yellow in listings)

## Screenshots
screenshot 2018-12-19 at 14 59 37

### PR DESCRIPTION
**Pull request by rvillablanca**
This PR fix #5566 so now a due date that is today won't be considered overdue, this applies to milestones and issues.

### CODE REVIEW

File: models/issue.go
Comment (Bwko) [inline models/issue.go]: It's possible to replace this complete IsOverdue method with this:
```go
deadline := issue.DeadlineUnix.AsTime().UTC().Round(24*time.Hour) // Round to end of day
return time.Now().UTC().After(deadline)
```

File: models/issue_milestone.go
Comment (Bwko) [inline models/issue_milestone.go]:
```go
deadline := m.DeadlineUnix.AsTime().UTC().Round(24*time.Hour) // Round to end of day
if time.Now().UTC().After(deadline) {
	m.IsOverdue = true
}
```

**Review by Bwko — CHANGES_REQUESTED** [review]

**Review by Bwko — APPROVED** [review]

File: models/issue.go
Comment (kolaente) [inline models/issue.go]: I'd rather change the behaviour when saving a due date to save it on 23:59:59 instead of 00:00 rather than changing this when showing the due date.

**Review by kolaente — CHANGES_REQUESTED** [review]

File: models/issue.go
Comment (lunny) [reply models/issue.go]:
```Go
return util.TimeStampNow() >= issue.DeadlineUnix + util.TimeStamp(time.Hour * 24 / time.Second)
```

Comment (Bwko) [reply models/issue.go]: I agree with kolaente

Comment (rvillablanca) [reply models/issue.go]: Ok thinking about the solution It would be like this

```go
if form.Deadline != nil && !form.Deadline.IsZero() {
	t := form.Deadline
	deadLine := time.Date(t.Year(), t.Month(), t.Day(), 23, 59, 59, 0, t.Location())
	form.Deadline = &deadLine
	deadlineUnix = util.TimeStamp(form.Deadline.Unix())
}
```
This is an example for the issue deadline case. What do you think @lunny, @kolaente, @Bwko, @adelowo  ? Maybe this code could be simplified.

Comment (kolaente) [reply models/issue.go]: I'd only set it to 23:59:59 if it is set to 00:00:00. Maybe someone wants to use it in some kind of frontend with exact deadlines. Or maybe don't modify it at all on the backend and only in gitea's frontend?

Comment (lunny) [reply models/issue.go]: @rvillablanca I don't against @kolaente 's idea, but that you have to consider old data migration.

Comment (adelowo) [reply models/issue.go]: @kolaente I don't think the frontend currently supports ___exact deadlines___.

Comment (kolaente) [reply models/issue.go]: @adelowo Ours doesn't, but the api does.

Comment (lafriks) [reply models/issue.go]: I don't think we should support exact deadlines (with time)

File: routers/api/v1/repo/issue.go
Comment (kolaente) [inline routers/api/v1/repo/issue.go:418]: I'd rather change the request Gitea makes to set the deadline instead of changing it here. Or we should'nt accept full datetime objects here anymore.

**Review by kolaente — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/issue.go
Comment (lafriks) [reply routers/api/v1/repo/issue.go:418]: I yet to see project where deadlines are planned by exact time. We should accept only date imho

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
@kolaente need your approval.

File: routers/api/v1/repo/issue.go
Comment (kolaente) [reply routers/api/v1/repo/issue.go:418]: @lafriks If we do, we shouldn't accept full datetime objects, only dates. I can see how this can lead to confiusion.

Comment (techknowlogick) [reply routers/api/v1/repo/issue.go:418]: I agree that we shouldn't accept full datetime objects, however as we currently do I think that change should be in a new PR marked as breaking, perhaps against the 1.8 release.

@kolaente as this is a blocking PR for 1.7 release, would you be able to either accept changes as is, or give permission to dismiss your review?

Comment (kolaente) [reply routers/api/v1/repo/issue.go:418]: @techknowlogick ok, but can we at least add a notice in the swaggerdocs stating Gitea will ignore any passed time with the date and only use the date?

Comment (techknowlogick) [reply routers/api/v1/repo/issue.go:418]: @kolaente that sounds reasonable. ~I will create a PR for that in the SDK repo to add that to swaggerdocs.~ Edit: nvm, best to create that in this PR, and I will do that now.

**Review by kolaente — APPROVED** [review]
Looks good! I think the CI failed though...

**Pull request merged**
